### DISCUSSION TITLE
How to represent sequences that can be joined in any order in a tree-sitter grammar file

### DISCUSSION THREAD

**Discussion by bytesnail** [tree-sitter/tree-sitter#2416] [Q&A] [answered]
Is it wrong to write it this way ?
``` javascript
seq(rule1, rule2, rule3) 
```

I want to match the following sentences:
```
rule1, rule2, rule3
rule1, rule3, rule2
rule2, rule1, rule3
rule2, rule3, rule1
rule3, rule1, rule2
rule3, rule2, rule1
```

I really hope to get anyone's guidance, thank you very much!

---------------------------- edit --------------------------------
I'm trying to convert EBNF rules like the following to tree-sitter grammar, pay special attention to the comment "These stmts can appear in any order"

```
include-stmt        = include-keyword sep identifier-arg-str optsep
                      (";" /
                       "{" stmtsep
                           ;; these stmts can appear in any order
                           [revision-date-stmt]
                           [description-stmt]
                           [reference-stmt]
                       "}") stmtsep
```

**Comment by ahlinc** [comment]
Something like:
```js
input: $ => seq(
  $.triple
  repeat(seq(/\n/, $.triple)),
),
triple: $ => seq(
  $.identifier, ",",
  $.identifier, ",",
  $.identifier,
),
identifier: $ => /[a-z0-9]+/,
```

**Reply by bytesnail** [reply]
I think there is some misunderstanding here. It is my bad, because I didn't make the issue clear.

**Reply by bytesnail** [reply]
The above string does not refer to the actual string rule1/2/3, but to any string that matches rule 1/2/3

**Reply by bytesnail** [reply]
The 6 lines of example string given are not a whole, but express the possibility of splicing in different orders in 6 of 3 rules. I wanted to write a rule that would express this splicing combination without a fixed order

**Reply by bytesnail** [reply]
English is not my native language, and I'm sorry if something is wrong

**Reply by sogaiu** [reply]
Is it possible that `choice` might help in this situation?

FWIW, [this section of the docs](https://tree-sitter.github.io/tree-sitter/creating-parsers#the-grammar-dsl) has related info:

> Alternatives : choice(rule1, rule2, ...) - This function creates a rule that matches one of a set of possible rules. The order of the arguments does not matter. This is analogous to the | (pipe) operator in EBNF notation.

**Reply by bytesnail** [reply]
You mean choice(rule1, rule2, rule3) repeated 3 times, like seq(choice(rule1, rule2, rule3), choice(rule1, rule2, rule3), choice(rule1, rule2, rule3)) ?

This may not work, it will match something like 'rule1 rule1 rule2' or 'rule1 rule1 rule1'. This creates some duplication.

I guess the EBNF rule wants to match these different expressions that each appears once or zero times (in any order), but there should not be consecutive repetitions of either statement.

**Reply by amaanq** [reply, marked as answer]
you could write out every possible combination, like so:
```scm
choice(
  seq(x, y, z),
  seq(z, y, x),
  ..etc
)
```

**Reply by bytesnail** [reply]
I already have some ideas, I need to use seq and choice to wrap a new function, inside the function will first the received rule list parameters are fully arranged, to get a total of n! sort order, use seq in any sort order for splicing, and wrap it with a choice between different permutation orders, so that as a whole, I think we get a new "seq" that can be spliced in any order.

**Reply by bytesnail** [reply]
bingo! You and me think the same thing.

**Reply by bytesnail** [reply]
Is it possible that this will become a built-in helper function in the future, a new "seq" that allows elements to be spliced in any order?

**Reply by amaanq** [reply]
No not really, imo tree-sitter is okay (and better off) being looser w/ what syntax it accepts for a couple reasons

1) Reduce state count/complexity, instead of NxN possible nodes, you have a repetition of 3 possible nodes
2) Better error recovery, someone writing "bad" code will have it parse okay as long as the constructs "look" okay (this is more subjective)
